domain/src/test/java/org/oppia/android/domain/spotlight/SpotlightStateControllerTest.kt

```diff
@@ -0,0 +1,388 @@
+import org.oppia.android.testing.TestLogReportingModule
+import org.oppia.android.testing.data.DataProviderTestMonitor
+import org.oppia.android.testing.environment.TestEnvironmentConfig
+import org.oppia.android.testing.robolectric.RobolectricModule
+import org.oppia.android.testing.threading.TestDispatcherModule
+import org.oppia.android.testing.time.FakeOppiaClockModule
+import org.oppia.android.util.caching.AssetModule
+import org.oppia.android.util.caching.CacheAssetsLocally
+import org.oppia.android.util.caching.LoadLessonProtosFromAssets
+import org.oppia.android.util.caching.TopicListToCache
+import org.oppia.android.util.data.DataProvidersInjector
+import org.oppia.android.util.data.DataProvidersInjectorProvider
+import org.oppia.android.util.locale.LocaleProdModule
+import org.oppia.android.util.logging.EnableConsoleLog
+import org.oppia.android.util.logging.EnableFileLog
+import org.oppia.android.util.logging.GlobalLogLevel
+import org.oppia.android.util.logging.LogLevel
+import org.oppia.android.util.logging.SyncStatusModule
+import org.oppia.android.util.networking.NetworkConnectionUtilDebugModule
+import org.oppia.android.util.platformparameter.LearnerStudyAnalytics
+import org.oppia.android.util.platformparameter.PlatformParameterValue
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Suppress("SameParameterValue", "FunctionName")
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(application = SpotlightStateControllerTest.TestApplication::class)
+class SpotlightStateControllerTest {
+
+  @Inject
+  lateinit var spotlightStateController: SpotlightStateController
+
+  @Inject
+  lateinit var dataProviderTestMonitor: DataProviderTestMonitor.Factory
+
+  private val profileId = ProfileId.newBuilder().setInternalId(0).build()
+  private val profileId1 = ProfileId.newBuilder().setInternalId(1).build()
+
+  @Before
+  fun setUp() {
+    setUpTestApplicationComponent()
+  }
+
+  @Test
+  fun testMarkSpotlightState_validFeature_notYetMarked_returnsSuccess() {
+    val markSpotlightProvider =
+      spotlightStateController.markSpotlightViewed(profileId, FIRST_CHAPTER)
+    dataProviderTestMonitor.waitForNextSuccessfulResult(markSpotlightProvider)
+  }
+
+  @Test
+  fun testMarkSpotlightState_validFeature_alreadyMarked_returnsSuccess() {
+    markSpotlightSeen(FIRST_CHAPTER)
+    val markSpotlightProvider =
+      spotlightStateController.markSpotlightViewed(profileId, FIRST_CHAPTER)
+    dataProviderTestMonitor.waitForNextSuccessfulResult(markSpotlightProvider)
+  }
+
+  @Test
+  fun testMarkSpotlightView_invalidFeature_returnsFailure() {
+    val invalidFeature = Spotlight.FeatureCase.FEATURE_NOT_SET
+    val markSpotlightProvider =
+      spotlightStateController.markSpotlightViewed(profileId, invalidFeature)
+    dataProviderTestMonitor.waitForNextFailureResult(markSpotlightProvider)
```

Comment:
This should verify that the correct failure is being returned (e.g. by checking the instance type & exception cause method).

Reply:
I'm facing a problem doing this. Pushed my latest code -- PTAL;  P.S. the markSpotlightSeen() function has a try catch block at the end.

Reply:
I left a new comment in the latest review pass--I think you might be checking it incorrectly (see that comment--we shouldn't be using ``assertThrows`` here). I also suggest looking at other tests that use the monitor for testing failures; they'll probably be really helpful as references.